File: 20_engine_editor_00_core_and_application_layer/src/Lua/renderer/material.lua
require("lua_extension")
require("cpp_class")

--- @class Material 材质
Material=class("Material",CppClass)

function Material:ctor()
    Material.super.ctor(self)
end

--- 实例化C++ Class
function Material:InitCppClass()
    self.cpp_class_instance_=Cpp.Material()
end

--- 加载Material文件并解析
function Material:Parse(material_path)
    self.cpp_class_instance_:Parse(material_path)
end

--- 上传int值
--- @param shader_property_name string @shader属性名
--- @param value number @值
function Material:SetUniform1i(shader_property_name,value)
    self.cpp_class_instance_:SetUniform1i(shader_property_name,value)
end

--- 上传float值
--- @param shader_property_name string @shader属性名
--- @param value number @值
function Material:SetUniform1f(shader_property_name,value)
    self.cpp_class_instance_:SetUniform1f(shader_property_name,value)
end

--- 上传vec3值
--- @param shader_property_name string @shader属性名
--- @param value glm.vec3 @值
function Material:SetUniform3f(shader_property_name,value)
    self.cpp_class_instance_:SetUniform3f(shader_property_name,value)
end

--- 设置纹理
--- @param property
--- @param texture2D
function Material:SetTexture(property,texture2D)
    self.cpp_class_instance_:SetTexture(property,texture2D)
end

function Material:shader()
    return self.cpp_class_instance_:shader()
end

function Material:textures()
    return self.cpp_class_instance_:textures()
end

File: 19_classic_lighting_05_ubo/src/Engine/Renderer/Material.cpp
#include "Engine/Renderer/Material.hpp"

#include <fstream>
#include <iostream>

#include "Engine/Renderer/Shader.hpp"
#include "Engine/Renderer/Texture2D.hpp"
#include "Engine/Utils/Application.hpp"
#include "glad/glad.h"
#include "rapidxml/rapidxml.hpp"
#include "rapidxml/rapidxml_print.hpp"
#include "rapidxml/rapidxml_utils.hpp"

Material::Material() {}

Material::~Material() {}

void Material::Parse(const std::string& material_path) {
  // 解析xml
  rapidxml::file<> xml_file((Application::data_path() + material_path).c_str());
  rapidxml::xml_document<> document;
  document.parse<0>(xml_file.data());

  // 根节点
  rapidxml::xml_node<>* material_node = document.first_node("material");
  if (material_node == nullptr) {
    return;
  }

  rapidxml::xml_attribute<>* material_shader_attribute =
      material_node->first_attribute("shader");
  if (material_shader_attribute == nullptr) {
    return;
  }
  shader_ = Shader::Find(material_shader_attribute->value());

  // 解析Texture
  rapidxml::xml_node<>* material_texture_node =
      material_node->first_node("texture");
  while (material_texture_node != nullptr) {
    rapidxml::xml_attribute<>* texture_name_attribute =
        material_texture_node->first_attribute("name");
    if (texture_name_attribute == nullptr) {
      return;
    }

    rapidxml::xml_attribute<>* texture_image_attribute =
        material_texture_node->first_attribute("image");
    if (texture_image_attribute == nullptr) {
      return;
    }

    std::string shader_property_name = texture_name_attribute->value();
    std::string image_path = texture_image_attribute->value();
    textures_.push_back(std::make_pair(texture_name_attribute->value(),
                                       Texture2D::LoadFromFile(image_path)));

    material_texture_node = material_texture_node->next_sibling("texture");
  }
}

void Material::SetUniform1i(const std::string& shader_property_name, int value) {
    uniform_1i_map_[shader_property_name]= value;
}

void Material::SetUniform1f(const std::string& shader_property_name, float value) {
    uniform_1f_map_[shader_property_name]=  value;
}

void Material::SetUniform3f(const std::string& shader_property_name,glm::vec3& value){
    uniform_3f_map_[shader_property_name]= value;
}

void Material::SetTexture(const std::string& property, Texture2D *texture2D) {
    for (auto& pair : textures_){
        if(pair.first==property){
            delete(pair.second);
            pair.second=texture2D;
            break;
        }
    }
}